cli/cmd/root.go

```diff
@@ -31,39 +31,47 @@ var rootCmd = &cobra.Command{
 
 // Execute adds all child commands to the root command and sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
-func Execute(ctx context.Context, ver version.Version) {
+func Execute(ctx context.Context, ver config.Version) {
 	err := runCmd(ctx, ver)
 	if err != nil {
 		os.Exit(1)
 	}
 }
 
-func runCmd(ctx context.Context, ver version.Version) error {
+func runCmd(ctx context.Context, ver config.Version) error {
+	// Cobra config
 	rootCmd.Version = ver.String()
+	rootCmd.PersistentFlags().BoolP("help", "h", false, "Print usage") // Overrides message for help
+	rootCmd.Flags().BoolP("version", "v", false, "Show rill version")  // Adds option to get version by passing --version or -v
+
+	// Build CLI config
+	cfg := &config.Config{
+		Version: ver,
+	}
 
-	rootCmd.AddCommand(initialize.InitCmd(ver))
-	rootCmd.AddCommand(start.StartCmd(ver))
-	rootCmd.AddCommand(build.BuildCmd(ver))
-	rootCmd.AddCommand(source.SourceCmd(ver))
-	rootCmd.AddCommand(admin.AdminCmd(ver))
-	rootCmd.AddCommand(runtime.RuntimeCmd(ver))
+	// Add sub-commands
+	rootCmd.AddCommand(initialize.InitCmd(cfg))
+	rootCmd.AddCommand(start.StartCmd(cfg))
+	rootCmd.AddCommand(build.BuildCmd(cfg))
+	rootCmd.AddCommand(source.SourceCmd(cfg))
+	rootCmd.AddCommand(admin.AdminCmd(cfg))
+	rootCmd.AddCommand(runtime.RuntimeCmd(cfg))
 	rootCmd.AddCommand(docs.DocsCmd())
 	rootCmd.AddCommand(completionCmd)
 	rootCmd.AddCommand(versioncmd.VersionCmd())
-	rootCmd.AddCommand(auth.AuthCmd(ver))
-	rootCmd.AddCommand(org.OrgCmd(ver))
-	rootCmd.AddCommand(project.ProjectCmd(ver))
 
-	// Here you will define your flags and configuration settings.
-	// Cobra supports persistent flags, which, if defined here,
-	// will be global for your application.
-	// rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.cli.yaml)")
-	rootCmd.PersistentFlags().BoolP("help", "h", false, "Print usage") // Overrides message for help
-
-	// Cobra also supports local flags, which will only run
-	// when this action is called directly.
-	// rootCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
-	rootCmd.Flags().BoolP("version", "v", false, "Show rill version")
+	// Add sub-commands for admin
+	// (This allows us to add persistent flags that apply only to the admin-related commands.)
+	adminCmds := []*cobra.Command{
+		auth.AuthCmd(cfg),
+		org.OrgCmd(cfg),
+		project.ProjectCmd(cfg),
+	}
+	for _, cmd := range adminCmds {
+		cmd.PersistentFlags().StringVar(&cfg.AdminURL, "api-url", "https://admin.rilldata.com", "Base URL for the admin API")
+		cmd.PersistentFlags().StringVar(&cfg.AdminToken, "api-token", "", "Token for authenticating with the admin API")
+		rootCmd.AddCommand(cmd)
+	}
```

Comment:
Can't move this as part of respective commands?

Reply:
I think it's better to define them once like this at the root, since they would otherwise get replicated quite many places as we add more admin-related commands.

I even considered adding them directly to the `root` command, but it felt wrong since they're not used for `start`, `init`, etc.